src/solana/SolanaEngine.ts

```diff
@@ -157,48 +156,39 @@ export class SolanaEngine extends CurrencyEngine<
             this.base58PublicKey,
             { commitment: this.networkInfo.commitment }
           ]
-        }
-      ]
-
-      const allTokenIds = [...Object.keys(this.allTokensMap)]
-      const pubkey = new PublicKey(this.base58PublicKey)
-      const tokenProgramId = new PublicKey(this.networkInfo.tokenPublicKey)
-      const associatedTokenProgramId = new PublicKey(
-        this.networkInfo.associatedTokenPublicKey
-      )
-      for (const tokenId of allTokenIds) {
-        const associatedTokenPubkey = getAssociatedTokenAddressSync(
-          new PublicKey(tokenId),
-          pubkey,
-          false,
-          tokenProgramId,
-          associatedTokenProgramId
-        )
-        requests.push({
-          method: 'getTokenAccountBalance',
+        },
+        {
+          method: 'getTokenAccountsByOwner',
           params: [
-            associatedTokenPubkey.toBase58(),
-            {
-              commitment: this.networkInfo.commitment,
-              encoding: 'jsonParsed'
-            }
+            this.base58PublicKey,
+            { programId: this.networkInfo.tokenPublicKey },
+            { encoding: 'jsonParsed' }
           ]
-        })
-      }
+        }
+      ]
 
       const balances: any = await this.fetchRpcBulk(requests)
 
-      const [mainnetBal, ...tokenBals]: [AccountBalance, TokenAmount[]] =
+      const [mainnetBal, tokenBalsRaw]: [AccountBalance, TokenAmount[]] =
         balances
+
       const balance = asAccountBalance(mainnetBal)
       this.updateBalance(this.chainCode, balance.result.value.toString())
 
+      const tokenBalances = asTokenBalances(tokenBalsRaw).result.value
+      const tokenBalMap = tokenBalances.reduce(
+        (acc: { [key: string]: string }, tokenBal) => {
+          acc[tokenBal.account.data.parsed.info.mint] =
+            tokenBal.account.data.parsed.info.tokenAmount.amount
+          return acc
+        },
+        {}
+      )
+
+      const allTokenIds = [...Object.keys(this.allTokensMap)]
       const detectedTokenIds: string[] = []
-      for (const [i, tokenId] of allTokenIds.entries()) {
-        const tokenBal = asMaybe(asTokenBalance)(tokenBals[i])
-        // empty token addresses return an error "Invalid param: could not find account".
-        // If there was an actual error with the request it would have thrown already
-        const balance = tokenBal?.result?.value?.amount ?? '0'
+      for (const tokenId of allTokenIds) {
```

Comment:
Faster & easier to read if you get rid of the `allTokenIds` variable: `for (const tokenId of Object.keys(this.allTokensMap))`.

Also, why are we looping over every token, including disabled ones? Is this for auto-detect tokens? Otherwise, this.enabledTokenIds seems better.

Reply:
Correct it's for auto-detection